Allow configuration warnings to refer to a property (reverted)

RedMser opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/550

Configuration warnings changed from PackedStringArray to an Array that accepts a mix of Strings and Dictionaries `{ message: String, property?: String }`.
If `property` is set, then show a clickable warning icon on property in the inspector.
Strings in this array are converted to an equivalent Dictionary with `message` filled.

This PR also lays the groundwork for further configuration warning enhancements like https://github.com/godotengine/godot-proposals/issues/2519

As an example, the PointLight2D node was updated to use this system.

https://user-images.githubusercontent.com/5117197/200654466-370da225-17ea-4d90-85c5-04b5bcd7bc49.mp4

---

The PR is split into two commits for easier reviewing. The second commit only fixes the signature of the `get_configuration_warnings` method for all classes using it.

---

**TODO:**

- [x] Change warning icon based on how many warnings the property has. Probably have to move the property name filtering logic out of the `as_string` method.
- [ ] Would be nice to have a tooltip when hovering the icon, but since other icons don't have this yet, it can be done in a follow-up PR.
- [x] It seems to be necessary to introduce a compatibility method for GDExtension here.
  - [ ] No idea how to test it, so someone should verify what it's doing is correct. Especially since there's also a virtual method involved here, whose signature changed but didn't get any compat bind (I followed what was done in AnimationMixer.compat.inc)!

Mickeon commented:
Wow... what work. My only concern is that it's not just a breaking change after 4.0, but it'd make the virtual method ever so laborious to use for simpler cases. Having to create `{ message: "Something is wrong etc. etc."}`, that is. 

I can assume that my concerns are exaggerated, as not everyone casually makes use of the method, but it makes me wonder if it could be an Array of Variant, where each one of them can either be a String or a Dictionary. This way the previous, simplistic behavior is also kept.

RedMser commented:
@Mickeon I totally agree. Since there is no concrete proposal behind this, I implemented it in the most straightforward way I could think of. But for the sake of compatibility and convenience, I'd agree that treating plain strings just like before would be nice.

Would like some more opinions on this before I try implementing anything more, e.g. if this is the kind of API breakage that's also acceptable for minor versions.

MewPurPur commented:
I think it's a bit weird duplicating config warnings like this. It gives off a feeling that two things are wrong, instead of just one.

If I gather correctly, this can be implemented in a not compat-breaking way, by having separate "node configuration" and "property configuration" warnings? Here's my idea:
- Keep config warnings as is.
- Have property warnings point to a property with a warning message.
- Property warnings update the config warning in standardized ways, i.e. "[property_name] of this [node_name] has not been set up properly."

RedMser committed:
Allow configuration warnings to refer to a property

This is used by the inspector so it can show a warning icon on
a specific property.

RedMser committed:
Update Node::get_configuration_warnings signature

RedMser commented:
Updated so that `get_configuration_warnings` may be a mixed array of Strings and Dictionaries. This keeps compatibility with existing C++ and GDScript code, and keeps an easy way to write simple configuration warnings.

Also for easier review, I split the changes into two commits, since one is just a large batch of signature changes which do not need reviewing.

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -1008,6 +1094,7 @@ void EditorProperty::_bind_methods() {
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "draw_warning"), "set_draw_warning", "is_draw_warning");
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "keying"), "set_keying", "is_keying");
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "deletable"), "set_deletable", "is_deletable");
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "configuration_warning"), "set_configuration_warning", "get_configuration_warning");
```
```suggestion
	ADD_PROPERTY(PropertyInfo(Variant::STRING, "configuration_warning"), "set_configuration_warning", "get_configuration_warning");
```

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -1037,6 +1124,10 @@ EditorProperty::EditorProperty() {
 	menu = nullptr;
 	set_process_shortcut_input(true);
+
+	warning_dialog = memnew(AcceptDialog);
```
This should be created only on demand.

RedMser replied:
I took this code from `SceneTreeEditor`, so I will also fix it there then.

RedMser replied:
Ah right, I see what you mean now. I misremembered, thinking the dialog was in the inspector and not per property for some reason.

KoBeWi commented on scene/main/node.cpp:
```diff
@@ -3017,30 +3017,78 @@ void Node::clear_internal_tree_resource_paths() {
 	}
+}
 
+TypedArray<Dictionary> Node::get_configuration_warnings_as_dicts() const {
```
This is an internal code, so you could use `Vector<Dictionary>`.

RedMser commented on editor/editor_inspector.cpp:
```diff
@@ -721,6 +775,24 @@ void EditorProperty::gui_input(const Ref<InputEvent> &p_event) {
+			for (int i = 0; i < boundaries.size(); i += 2) {
+				const int start = boundaries[i];
+				const int end = boundaries[i + 1];
+				lines.append(configuration_warning.substr(start, end - start + 1));
```
Once #79054 is merged, the `+ 1` has to be removed from here.

KoBeWi reviewed (approved):
The implementation looks fine (aside from breaking compatibility; maybe there is a solution that could avoid it). There is nice flexibility with Dictionaries, which could help with #74357

>  Would be nice to have a tooltip when hovering the icon, but it doesn't seem like that's easily possible with these "inline" icons? Would like to hear suggestions on how to tackle this.

This can be implemented using `get_tooltip()` virtual method and `new_configuration_warning_hover`. But other icons don't have tooltips; it could be implemented for all of them in another PR.

Mickeon commented:
Yyyeaah, this one cannot be done now without breaking compatibility, unless it is possible for it not to.

RedMser commented:
Hi @Mickeon , your PR #87535 removed `Node::get_configuration_warnings_as_string` which I intended to use here.

Clicking on the :warning: button that appears next to a property, a dialog similar to the one from the scene tree would show up. It would filter to only show relevant warnings, so the function would get an optional property parameter.

Without the function, this would involve duplicating the logic involving the bullet points, indenting new lines etc.

Do you think it would be fine to re-introduce it here, now that there'd be multiple places that need it for different purposes? Or was there a strong reason against having that function around in the first place?

Mickeon commented:
> Do you think it would be fine to re-introduce it here, now that there'd be multiple places that need it for different purposes? Or was there a strong reason against having that function around in the first place?

The reason it was removed was because it was only used two times: The tooltip, and the popup. However, the way these two are displayed is **entirely** different, to the point where relying on that function's result was basically useless, as the strings had to be reformatted anyhow. It also was not encompassed with `TOOLS_ENABLED`, which included it in release builds for no reason (if the compiler is stupid).

With that said, I am not sure if it should be brought back. At the same time, this PR may override what was made in https://github.com/godotengine/godot/pull/87535, so you may as well... go all in?

AThousandShips commented on misc/extension_api_validation/4.2-stable.expected:
```diff
@@ -82,3 +82,9 @@ Validate extension JSON: Error: Field 'classes/GPUParticles3D/properties/process
 
 Property hints reordered to improve editor usability. The types allowed are still the same as before. No adjustments should be necessary.
+
+GH-68420
```
```suggestion

GH-68420
```

AThousandShips reviewed (approved):
Code and documentation looks good

akien-mga commented on editor/editor_inspector.cpp:
```diff
@@ -911,6 +985,14 @@ void EditorProperty::_update_pin_flags() {
 }
 
+void EditorProperty::_update_configuration_warnings() {
+	if (Node *node = Object::cast_to<Node>(object)) {
```
Currently we have a convention not to use this type of construct with an assignment in an `if` condition, so this should be:
```suggestion
	Node *node = Object::cast_to<Node>(object)
	if (node) {
```

akien-mga commented:
Thanks!

Zylann commented:
What should GDScript plugins do if they support engine versions prior this change?
```gdscript
func _get_configuration_warnings() -> PackedStringArray:
```

> Updated so that get_configuration_warnings may be a mixed array of Strings and Dictionaries. This keeps compatibility with existing C++ and GDScript code

GDScript doesn't have a preprocessor to make this conditionally implemented differently. If I write `-> Array`, it will break in versions prior this change.
That means I have to remove typing completely.

KoBeWi commented:
> That means I have to remove typing completely.

Yes, that's the solution if you want to support older versions.

aaronfranke commented on scene/main/node.h:
```diff
@@ -636,7 +637,11 @@ class Node : public Object {
 
 	_FORCE_INLINE_ Viewport *get_viewport() const { return data.viewport; }
 
-	virtual PackedStringArray get_configuration_warnings() const;
+	virtual Array get_configuration_warnings() const;
+	Dictionary configuration_warning_to_dict(const Variant &p_warning) const;
+	Vector<Dictionary> get_configuration_warnings_as_dicts() const;
+	Vector<Dictionary> get_configuration_warnings_of_property(const String &p_property = String()) const;
+	PackedStringArray get_configuration_warnings_as_strings(bool p_wrap_lines, const String &p_property = String()) const;
```
In https://github.com/godotengine/godot/pull/81892 I proposed removing the String conversion method from `Node`, as it's only used by the editor. This PR is going in the opposite direction, adding more formatting methods to `Node`.

IMO, only `virtual Array get_configuration_warnings() const;` should exist in `Node`. The rest are editor-specific.

RedMser replied:
That is very fair. Since they are not bound methods (and shouldn't be), it's not a problem to move them elsewhere as a refactor further down the line.

I just don't know enough about organization in C++ to really know a good place for them, or if you can just create a `editor_configuration_warnings.cpp` file for this and it'll magically work?

Mickeon replied:
Note that it was also removed in https://github.com/godotengine/godot/pull/87535 for similar reasons.

dsnopek commented on scene/main/node.h:
```diff
@@ -320,7 +321,7 @@ class Node : public Object {
 	GDVIRTUAL0(_enter_tree)
 	GDVIRTUAL0(_exit_tree)
 	GDVIRTUAL0(_ready)
-	GDVIRTUAL0RC(Vector<String>, _get_configuration_warnings)
+	GDVIRTUAL0RC(Array, _get_configuration_warnings)
```
This breaks binary compatibility for GDExtension. Any GDExtension that implements `_get_configuration_warnings()` from before this change will probably segfault. Does anyone know if there's an issue discussing this yet? If not, I can make one

dsnopek replied:
I just created issue https://github.com/godotengine/godot/issues/88379
